app/pagestack.go

```diff
@@ -32,23 +32,25 @@ func (pageStack *PageStack) Top() Page {
 	return nil
 }
 
+// NavigateFromCurrentPage triggers the OnNavigatedFrom() method of the current
+// page.
+func (pageStack *PageStack) NavigateFromCurrentPage() {
+	page := pageStack.Top()
+	if page != nil {
+		page.OnNavigatedFrom()
+	}
+}
+
 // Push pushes the specified page to the top of the stack, removing all other
-// instances of the same page from the stack. An about-to-display signal is sent
-// to the new page via newPage.OnNavigatedTo() while page.OnNavigatedFrom() is
-// called on the current page to signal that the current page is no longer the
-// displayed page.
-func (pageStack *PageStack) Push(newPage Page, navigator PageNavigator) bool {
+// instances of the same page from the stack.
+func (pageStack *PageStack) Push(newPage Page) bool {
+	if currentPage := pageStack.Top(); currentPage != nil && currentPage.ID() == newPage.ID() {
+		return false
+	}
```

Comment:
What's the motivation for the change to this method? This change introduces a potential issue. If the current page has the same ID as the new page, nothing is done by this method (which was the behaviour before this change), but the problem is, before this method was called, `NavigateFromCurrentPage()` was called. And so the current page would basically be useless now even though it remains on the top of the stack.

Discuss the motivation for the additional changes in this PR before making further changes otherwise revert to the previous code to avoid regressions.

Reply:
Dang. NavigateFromCurrentPage should be after the paste has been pushed and Push returned true.

I’m suggesting to reduce pageStack.Push responsibility because now that we have a start page when creating a master page, only callers can accurately tell when the new page is ready for display.

Reply:
> only callers can accurately tell when the new page is ready for display

Don't understand this. Why is this so?

Reply:
Now use Push in NewMasterPage and in Display. NewMasterPage cares only about preparing the page(I.e adding it to the page stack and binding the navigator to the start page) while Display does everything including notifying the page it is read for display by calling the OnNavigateTo method on a new page.

We will end up calling the top page OnNavigateTo method twice, If we continue to allow pageStack.Push to notify pages that they are ready (the first would be in pageStack.Push and the second would be when the MasterPage OnNavigatedTo Method is called).

Calling the OnNavigateTo method of a page twice in a roll (without calling OnNavigateFrom) is not without consequence, one of which is the fact page notifiers will be registered twice.

Reply:
I'm still confused.

Reply:
Typo

Reply:
Please see how NewMasterPage is used in this PR and masterpage.Display too.